app/models/concerns/orchestration/dhcp.rb

```diff
@@ -96,7 +96,15 @@ def boot_server
   def build_dhcp_record(record_mac)
     raise ::Foreman::Exception.new(N_("DHCP not supported for this NIC")) unless dhcp?
     record_attrs = dhcp_attrs(record_mac)
-    record_type = (provision? && jumpstart?) ? Net::DHCP::SparcRecord : Net::DHCP::Record
+    record_type = Net::DHCP::Record
+
+    if provision?
```

Comment:
I would extract that to a method or service class.

Reply:
Probably with switch-case for now.

Reply:
Hi,
Would something like this be okay for you:
```
  def get_record_type()
    record_type = Net::DHCP::Record
    if provision?
      if jumpstart?
        record_type = Net::DHCP::SparcRecord
      elsif operatingsystem.respond_to?(:pxe_type) && operatingsystem.pxe_type == "ZTP" && operatingsystem.respond_to?(:ztp_arguments)
        record_type = Net::DHCP::ZTPRecord
      end
    end
  end

  def build_dhcp_record(record_mac)
    raise ::Foreman::Exception.new(N_("DHCP not supported for this NIC")) unless dhcp?
    record_attrs = dhcp_attrs(record_mac)
    record_type = get_record_type

    handle_validation_errors do
      record_type.new(record_attrs)
    end
  end
```

Reply:
Something like this:

```ruby
  def get_record_type
    if provision?
      if jumpstart?
        return Net::DHCP::SparcRecord
      elsif operatingsystem.respond_to?(:pxe_type) && operatingsystem.pxe_type == "ZTP" && operatingsystem.respond_to?(:ztp_arguments)
        return Net::DHCP::ZTPRecord
      end
    end
    Net::DHCP::Record
  end
```

Reply:
On second thought: Even better would be to implement a `dhcp_record_type` method on the OS, `operatingsystem.dhcp_record_type`.

Something like this (needs more work, just to show the idea):
In base:

```ruby
def dhcp_record_type
  Net::DHCP::Record
end
```

For Sparc:
```ruby
def dhcp_record_type
  Net::DHCP:: SparcRecord
end
```

For ZTP:

```ruby
def dhcp_record_type
  return Net::DHCP::ZTPRecord if pxe_type == "ZTP"
  super
end
```

Reply:
Okay, will implement the latter, looks cleaner to me :-)

Reply:
Yup good idea.